submissions/2024-12-12-OpenFF-Sage-2.0.0-Training-Optimization-Dataset-v1.0/generate-combined-dataset.py

```diff
@@ -0,0 +1,149 @@
+
+import numpy as np
+import requests
+import logging
+logging.getLogger("openff").setLevel(logging.ERROR)
+
+from openff.qcsubmit.results import OptimizationResultCollection
+from openff.qcsubmit.datasets import OptimizationDataset
+from openff.qcsubmit.factories import OptimizationDatasetFactory
+from openff.qcsubmit.common_structures import MoleculeAttributes
+
+from openff.toolkit.utils import RDKitToolkitWrapper, ToolkitRegistry
+from openff.units import unit
+
+file = requests.get("https://raw.githubusercontent.com/openforcefield/openff-sage/refs/heads/main/data-set-curation/quantum-chemical/data-sets/1-2-0-opt-set-v3.json")
+filtered_and_combined = OptimizationResultCollection.parse_raw(file.content)
+rec_and_mol = filtered_and_combined.to_records()
+
+print('Number of results: ',filtered_and_combined.n_results,flush=True)
+print('Finished converting to records',flush = True)
+
+# Need to input initial molecules so that QCA will recognize the inputs as the same.
+initial_mols = [rec[0].initial_molecule for rec in rec_and_mol]
+
+dataset_factory1 = OptimizationDatasetFactory()
+provenance1 = dataset_factory1.provenance(ToolkitRegistry([RDKitToolkitWrapper]))
+
+dataset1 = OptimizationDataset(
+    dataset_name="OpenFF Sage 2.0.0 Training Optimization v1.0",
+    dataset_tagline="B3LYP-D3BJ/DZVP conformers applicable to drug-like molecules for OpenFF 2.0.0 Sage",
+    description=("A quantum chemical (QC) dataset curated to train OpenFF 2.0.0 Sage, "
+    "with reparametrized Lennard-Jones (LJ) and valence parameters, the latter "
+    "relevent to this dataset. This QC dataset with the OpenFF default level of "
+    "theory, B3LYP-D3BJ/DZVP, is used to benchmark Sage geometries and energetics. "
+    "These optimized conformer geometries where used in conjunction with the QC "
+    "dataset used to train one dimensional torsional profiles. This Generation 2 "
+    "dataset increases chemical diversity when compared to Generation 1, which are "
+    "of value to our industry partners. Large molecules (>20 heavy atoms) were also "
+    "included, including more flexible molecules and a greater degree of conformational "
+    "variation which provide intramolecular interactions."),
+    provenance=provenance1,
+)
+dataset1.metadata.submitter = "Jennifer A Clark"
+dataset1.metadata.long_description_url = (
+        "https://github.com/openforcefield/qca-dataset-submission/tree/master/"
+        "submissions/"
+        "2024-12-12-OpenFF-Sage-2.0.0-Training-Optimization-Dataset-v1.0"
+    )
+
+
+# Have to add records this way to avoid a round trip through the toolkit.
+records_by_cmiles= {}
+for record, molecule in rec_and_mol:
+    cmiles = molecule.to_smiles(isomeric=True, explicit_hydrogens=True, mapped=True)
```

Comment:
A bit nitpicky, but since this might establish a protocol for future tasks -- there's no guarantee that running a `to_smiles` will give you the same cmiles for the same molecule across different OpenEye/RDKit versions, or especially if you have one toolkit installed but another was used to generate the source datasets. It would be slightly more robust to use the exact cmiles in the dataset result. The easiest way to do this is probably by inspecting `filtered_and_combined.entries`, which is a dictionary where the key is the portal address and the value is a list of `OptimizationResult`s. Each `OptimizationResult` has an `id` and `cmiles` attribute that should correspond to the `record.id` of each record in `rec_and_mol`. Otherwise there may be other ways similar to what QCSubmit does when it downloads datasets (https://github.com/openforcefield/openff-qcsubmit/blob/d4e6b6986a58f5cf0184ba14dc4f7419e9978b67/openff/qcsubmit/results/results.py#L307-L353), but I haven't looked into those.

I'm not sure this actually matters very much; at the end of the day an OptimizationResultCollection or OptimizationDataset is a flat list of single-conformer results, and functionality like our ConformerRMSDFilter checks whether each record is a conformer of an existing molecule on-the-fly. However, it would be more ideal to keep the cmiles consistent between datasets just in case it does in the future.

Reply:
Oh thanks for this! I did notice that the CMILES changes depending on whether I use RDKit or OpenEye.

Reply:
Good catch Lily! Just in Jen's defense, I believe this code came straight from Lexie's NAGL2 combined dataset script. So I've also reviewed and approved it twice now, missing this subtlety. Also this is how `OptimizationResultCollection.create_basic_dataset` is implemented (except there `explicit_hydrogens=False, mapped=False` are passed), so Lexie probably got the code from there anyway.

This section of qcsubmit contains all of the locations you can check for a CMILES in an `OptimizationResultCollection`: https://github.com/openforcefield/openff-qcsubmit/blob/d4e6b6986a58f5cf0184ba14dc4f7419e9978b67/openff/qcsubmit/results/results.py#L483-L496. You can also be a bit clever with bool chaining:

```python
 def get_cmiles(entry):
     mol = entry.initial_molecule
     return (
         mol.identifiers.canonical_isomeric_explicit_hydrogen_mapped_smiles
         or mol.extras.get("canonical_isomeric_explicit_hydrogen_mapped_smiles")
         or entry.attributes.get(
             "canonical_isomeric_explicit_hydrogen_mapped_smiles"
         )
     )
```

But `entry` here has to be fetched from qcportal, so I agree that pulling the CMILES from the `OptimizationResultCollection` itself would be easier.

As a final observation, I think this line is not actually the problematic use of `to_smiles` in the script because this `cmiles` is only used to key a dict. `to_smiles` is called again on line 65, where it's actually stored in the dataset and would cause the kind of problem Lily is pointing out.

Reply:
Thanks, good catches Brent, yes line 65 is definitely where it matters!

You're more familiar with the code here than me -- if you think this will be problematic for `create_basic_dataset`, could you please raise an issue in qcsubmit?

@jaclark5 please ping me when you'd like a re-review -- I have notifications turned off for PR pushes so won't see any changes made.

Reply:
Just opened openforcefield/openff-qcsubmit#310 referencing this.